retrieval/target.go

```diff
@@ -267,6 +266,36 @@ func (t *Target) String() string {
 	return t.host()
 }
 
+// fingerprint returns an identifying hash for the target.
+func (t *Target) fingerprint() model.Fingerprint {
+	t.RLock()
+	defer t.RUnlock()
+
+	lset := t.Labels()
+
+	lset[model.MetricsPathLabel] = t.labels[model.MetricsPathLabel]
+	lset[model.AddressLabel] = t.labels[model.AddressLabel]
+	lset[model.SchemeLabel] = t.labels[model.SchemeLabel]
+
+	return lset.Fingerprint()
+}
+
+// offset returns the time until the next scrape cycle for the target.
+func (t *Target) offset(interval time.Duration) time.Duration {
+	now := time.Now().UnixNano()
+
+	var (
+		base   = now - (now % int64(interval))
+		offset = uint64(t.fingerprint()) % uint64(interval)
+		next   = base + int64(offset)
+	)
+
+	if next < now {
+		next += int64(interval)
+	}
+	return time.Duration(next - now)
```

Comment:
You add `now` in L288. Then you substract it here.

Just leave it out in both places and compare `next < 0` in L293. Should have the same effect.